{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "db47498f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Functions used in course\n",
    "#https://www.coursera.org/learn/bioinformatics/home/welcome\n",
    "\n",
    "### Part 1.2 funcion to count occurences of a pattern in given text.\n",
    "\n",
    "# return the number of occurences of \"Pattern\" in \"Text\"\n",
    "def PatternCount(Text, Pattern):\n",
    "    count = 0\n",
    "    for i in range(len(Text) - len(Pattern)+1):\n",
    "        if(Text[i:i+len(Pattern)] == Pattern):\n",
    "            count += 1\n",
    "    return count\n",
    "\n",
    "# return a dictionary with keys of all k-mers, and values of their occurences in Text\n",
    "def FrequencyMap(Text, k):\n",
    "    freq = {}\n",
    "    n = len(Text)\n",
    "    for i in range (n-k+1):\n",
    "        Pattern = Text[i:i+k]\n",
    "        freq[Pattern] = PatternCount(Text, Pattern)\n",
    "    return freq\n",
    "\n",
    "# return the k-mer with the highest frequency in Text\n",
    "def FrequentWords(Text, k):\n",
    "    words = []\n",
    "    freq = FrequencyMap(Text, k)\n",
    "    m = max(freq.values())\n",
    "    for key in freq:\n",
    "        if (freq[key]==m):\n",
    "            words.append(key)\n",
    "    return words\n",
    "\n",
    "# return the reverse of \"Pattern\"\n",
    "def Reverse(Pattern):\n",
    "    rev = \"\"\n",
    "    for char in Pattern:\n",
    "        rev = char + rev\n",
    "    return rev\n",
    "\n",
    "# return the complement of Pattern based on ATCG base pairing\n",
    "def Complement(Pattern):\n",
    "    comp = \"\"\n",
    "    for char in Pattern:\n",
    "        if char == \"A\":\n",
    "            comp = comp + \"T\"\n",
    "        if char == \"T\":\n",
    "            comp = comp + \"A\"\n",
    "        if char == \"C\":\n",
    "            comp = comp + \"G\"\n",
    "        if char == \"G\":\n",
    "            comp = comp + \"C\"\n",
    "    return comp\n",
    "\n",
    "# return the reverse complement\n",
    "def ReverseComplement(Pattern):\n",
    "    rev = \"\"\n",
    "    rev = Reverse(Pattern)\n",
    "    rev_comp = Complement(rev)\n",
    "    return rev_comp\n",
    "\n",
    "# return the positions in the genome where \"pattern\" is found\n",
    "def PatternMatching(Pattern, Genome):\n",
    "    positions = [] # output variable\n",
    "    for i in range(len(Genome) - len(Pattern)+1):\n",
    "        if(Genome[i:i+len(Pattern)] == Pattern):\n",
    "            positions.append(i)\n",
    "    return positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a66cfe7d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
